.github/workflows/run_tests.yml

```diff
@@ -0,0 +1,70 @@
+name: Run Tests
+
+on:
+  push:
+    branches: 
+      - '*'
+  pull_request:
+    branches:
+      - '*'
+
+jobs:
+  run_tests:
+    runs-on: ${{ matrix.config.os }}
+    if: "!contains(github.event.head_commit.message, 'ci skip')"
+    
+    strategy:
+      fail-fast: false
+      matrix:
+        config: 
+        - {name: 'current', os: ubuntu-latest, python: '3.8' }
+        - {name: 'prev', os: ubuntu-latest, python: '3.7' }
+        - {name: 'old', os: ubuntu-latest, python: '3.6' }
+
+    steps:
+    - uses: actions/checkout@v2
+    
+    - name: Install system dependencies
+      if: runner.os == 'Linux'
+      run: |
+        sudo apt-get update -qq
+        sudo apt-get install -y libhdf5-dev pandoc gfortran libblas-dev liblapack-dev libedit-dev llvm-dev
+    
+    - name: Install Singularity
+      uses: eWaterCycle/setup-singularity@v3
+      with:
+        singularity-version: 3.6.1
+    
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with: 
+        python-version: ${{ matrix.config.python }}
+
+    - name: Install package & dependencies
+      run: |
+        pip install -U .
+        python -c "import openproblems"
+        pip install -U .[test]
+        
+    - name: Linting
+      run: |
+        black . --check --diff --exclude=website
+    
+    - name: Build docker container # can snakemake do this?
+      run: |
+        for image in $(ls -1d docker/open* | sed 's#docker/##'); do
+          # only works when you have a DOCKER_PASSWORD variable defined, for now
+          # snakemake -j 1 docker/${image}/.docker_push
+          docker build -f docker/$image/Dockerfile -t singlecellopenproblems/$image .
```

Comment:
So there are two issues here:
1. Snakemake requires environment variables to be defined if included in env. We can probably just remove that line from the Snakefile.
2. Changes to the Dockerfiles will only be registered if they are pushed to Dockerhub. The alternative would be to have two pathways to the singularity image: a) docker build, docker push, singularity pull, singularity exec; or b) docker build, singularity build, singularity exec, with the decision between these two based on whether or not DOCKER_PASSWORD is defined.

Reply:
2. Third alternative: Is it not possible to just provide a DOCKER_EXEC in your snakemake as well, and letting the user choose whether to use singularity or snakemake? This would make the most sense to me, imho.

A fourth alternative is also possible, it _is_ possible to run singularity directory from an image built locally using docker:
```
$ docker build -f docker/openproblems/Dockerfile -t scop/base .
$ singularity exec docker-daemon:scop/base:latest ls -l /usr/src/
INFO:    Using cached SIF image
total 1
drwxr-xr-x. 8 root root 294 Oct 19 09:08 singlecellopenproblems
```

So it would be possible to use the following approach:
* If the docker command is available, and there is a 'singlecellopenproblems/openproblems' image available, use `singularity exec docker-daemon:singlecellopenproblems/openproblems:latest ...`
* Else, use `singularity exec docker://singlecellopenproblems/openproblems:latest ...`.

Reply:
I like this approach. I have some ideas on how to do this -- I'll give it a go.

Reply:
@rcannood `singularity exec docker-daemon:...` doesn't work for me and I can't find anything like this syntax in the docs. Can you point me to how to do this?

An alternative I just came up with would be to allow you to specify your own docker login as well as password

Reply:
@rcannood this should be resolved by #117.

Reply:
> singularity exec docker-daemon:... doesn't work for me and I can't find anything like this syntax in the docs. 

It's mentioned [here](https://sylabs.io/guides/3.0/user-guide/singularity_and_docker.html#locally-available-images-cached-by-docker) but only in the context of `singularity build`. For `singularity run` it also works, but I do need to take care to specify a tag, otherwise it doesn't work:

<details>
  <summary>singularity run docker-daemon:test:latest log</summary>
  
  ```
[rcannood@calpa SingleCellOpenProblems]$ singularity run docker-daemon:test:latest ls -l /usr/src/
INFO:    Converting OCI blobs to SIF format
WARNING: 'nodev' mount option set on /tmp, it could be a source of failure during build process
INFO:    Starting build...
Getting image source signatures
Copying blob b67d19e65ef6 done  
Copying blob 63ca78ed1eaf done  
Copying blob b5895172fcca done  
Copying blob c224480be65c done  
Copying blob f8a6bfa94a51 done  
Copying blob b879f0e06825 done  
Copying blob 0a836adbcb9c done  
Copying config 1e72a93671 done  
Writing manifest to image destination
Storing signatures
2020/10/27 06:38:11  info unpack layer: sha256:e881ec0ec79cc54b719a7a7a18821dd77029395ffde8b39485277cc945a30e4b
2020/10/27 06:38:11  warn xattr{etc/gshadow} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/10/27 06:38:11  warn xattr{/tmp/build-temp-239813469/rootfs/etc/gshadow} destination filesystem does not support xattrs, further warnings will be suppressed
2020/10/27 06:38:12  info unpack layer: sha256:51691093f3478e4b09afe29d333b9b634aac2c6780a9ea2f41f148088ae3c3ac
2020/10/27 06:38:12  warn xattr{var/cache/apt/archives/partial} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/10/27 06:38:12  warn xattr{/tmp/build-temp-239813469/rootfs/var/cache/apt/archives/partial} destination filesystem does not support xattrs, further warnings will be suppressed
2020/10/27 06:38:12  info unpack layer: sha256:940c6f20da96d24322cbf699f6389c7020d4257fd18b08a0e4eda47d775b6cac
2020/10/27 06:38:13  warn xattr{var/cache/apt/archives/partial} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/10/27 06:38:13  warn xattr{/tmp/build-temp-239813469/rootfs/var/cache/apt/archives/partial} destination filesystem does not support xattrs, further warnings will be suppressed
2020/10/27 06:38:13  info unpack layer: sha256:71d936d30a0f2d9a5fa8b3577ecc9eebc9c11d78f23c2aa846fe390251c5fb49
2020/10/27 06:38:13  info unpack layer: sha256:cde9d4e4ce8e7b754bb4afcf6ab9e06f57146f9dd20568b8eab88f907814045c
2020/10/27 06:38:13  warn xattr{var/cache/apt/archives/partial} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/10/27 06:38:13  warn xattr{/tmp/build-temp-239813469/rootfs/var/cache/apt/archives/partial} destination filesystem does not support xattrs, further warnings will be suppressed
2020/10/27 06:38:13  info unpack layer: sha256:01a2f15a3e4d78f19fbcd7004a5221963598b9fb6b17b76873672faa2476569b
2020/10/27 06:38:13  warn xattr{var/log/apt/term.log} ignoring ENOTSUP on setxattr "user.rootlesscontainers"
2020/10/27 06:38:13  warn xattr{/tmp/build-temp-239813469/rootfs/var/log/apt/term.log} destination filesystem does not support xattrs, further warnings will be suppressed
2020/10/27 06:38:13  info unpack layer: sha256:ebf410e9cbe3bce36e645651fba64a3e7ba7ed9d340f26f2ea1d41c54e2ee234
INFO:    Creating SIF file...
total 0
[rcannood@calpa SingleCellOpenProblems]$ singularity run docker-daemon:test ls -l /usr/src/
FATAL:   Unable to handle docker-daemon:test uri: failed to get checksum for docker-daemon:test: unable to parse image name docker-daemon:test: docker-daemon: reference test has neither a tag nor a digest
[rcannood@calpa SingleCellOpenProblems]$ 
```

</details>

> This should be resolved by #117.

I'll give it a shot!

Reply:
Seems to work fine. 

Running `nose2 ...`  results in some failed tests, but that's not due to Github Actions since it fails locally for me as well:

<details>
  <summary>nose2 errors</summary>

```
======================================================================
ERROR: openproblems.test.test_tools.transplant_class.<locals>.C (test_normalize_obsm_log_scran_pooling)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/parameterized/parameterized.py", line 530, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/home/runner/work/SingleCellOpenProblems/SingleCellOpenProblems/openproblems/test/test_tools.py", line 59, in test_normalize_obsm
    normalizer(adata, obsm=obsm)
  File "/home/runner/work/SingleCellOpenProblems/SingleCellOpenProblems/openproblems/tools/decorators.py", line 22, in normalize
    func(adata_temp, *args, **kwargs)
  File "/home/runner/work/SingleCellOpenProblems/SingleCellOpenProblems/openproblems/tools/normalize.py", line 13, in log_scran_pooling
    scprep.run.install_bioconductor("scran")
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/scprep/run/r_function.py", line 194, in install_bioconductor
    _install_bioconductor(**kwargs)
  File "<decorator-gen-30>", line 2, in __call__
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/scprep/utils.py", line 83, in _with_pkg
    return fun(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/scprep/run/r_function.py", line 141, in __call__
    robject = self.function(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/rpy2/robjects/functions.py", line 197, in __call__
    return (super(SignatureTranslatedFunction, self)
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/rpy2/robjects/functions.py", line 125, in __call__
    res = super(Function, self).__call__(*new_args, **new_kwargs)
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/rpy2/rinterface_lib/conversion.py", line 44, in _
    cdata = function(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/rpy2/rinterface.py", line 624, in __call__
    raise embedded.RRuntimeError(_rinterface._geterrmessage())
rpy2.rinterface_lib.embedded.RRuntimeError: Error in install.packages("BiocManager") : unable to install packages


======================================================================
ERROR: openproblems.test.test_methods.transplant_class.<locals>.C (test_method_openproblems_tasks_label_projection_pancreas_batch_logistic_regression_log_cpm)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/parameterized/parameterized.py", line 530, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/home/runner/work/SingleCellOpenProblems/SingleCellOpenProblems/openproblems/test/test_methods.py", line 161, in test_method
    run_image(
  File "/home/runner/work/SingleCellOpenProblems/SingleCellOpenProblems/openproblems/test/test_methods.py", line 144, in run_image
    utils.run(command)
  File "/home/runner/work/SingleCellOpenProblems/SingleCellOpenProblems/openproblems/test/utils.py", line 93, in run
    p = subprocess.run(command, shell=shell, stdout=subprocess.PIPE, stderr=stderr)
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/subprocess.py", line 489, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/subprocess.py", line 1637, in _execute_child
    self.pid = _posixsubprocess.fork_exec(
TypeError: expected str, bytes or os.PathLike object, not tuple
```
</details>